sns_module/src/main.rs

```diff
@@ -0,0 +1,207 @@
+use candid::{Nat, Principal};
+use ic_base_types::PrincipalId;
+use ic_cdk::{init, post_upgrade, query, update};
+use icp_ledger::{AccountIdentifier, Subaccount};
+use sns_module::memory::{
+    add_wtn_owed, decrease_wtn_owed, deposit_icp, get_principal_to_icp, get_principal_to_wtn_owed,
+};
+use sns_module::state::{read_state, replace_state, InitArg, State};
+use sns_module::{
+    balance_of, derive_staking, dispatch_tokens, is_distribution_available, is_swap_available,
+    transfer, Status, E8S, MIN_DEPOSIT_AMOUNT, NANOS,
+};
+
+fn main() {}
+
+#[init]
+fn init(args: InitArg) {
+    sns_module::memory::set_state(State::new(args.clone()));
+    let state = sns_module::memory::get_state().unwrap();
+    replace_state(state);
+}
+
+#[post_upgrade]
+fn post_upgrade() {
+    let state = sns_module::memory::get_state().unwrap();
+    replace_state(state);
+}
+
+#[query]
+fn get_status() -> Status {
+    let balances = get_principal_to_icp();
+    read_state(|s| Status {
+        participants: balances.len(),
+        total_icp_deposited: balances.iter().map(|(_, b)| b).sum(),
+        time_left: s.end_ts.saturating_sub(ic_cdk::api::time() / NANOS),
+        start_at: s.start_ts,
+        end_at: s.end_ts,
+        minimum_deposit_amount: MIN_DEPOSIT_AMOUNT,
+    })
+}
+
+#[query]
+fn get_principal_to_icp_deposited() -> Vec<(Principal, u64)> {
+    sns_module::memory::get_principal_to_icp()
+}
+
+#[query]
+fn get_icp_deposited(of: Principal) -> u64 {
+    sns_module::memory::get_icp_deposited(of)
+}
+
+#[query]
+fn get_principal_to_wtn_allocation() -> Vec<(Principal, u64)> {
+    get_principal_to_wtn_owed()
+}
+
+#[query]
+fn get_wtn_allocated(of: Principal) -> u64 {
+    sns_module::memory::get_wtn_owed(of)
+}
+
+#[update]
+fn get_icp_deposit_address(target: Principal) -> AccountIdentifier {
+    let subaccount = Subaccount(derive_staking(target));
+    let self_id: PrincipalId = ic_cdk::id().into();
+    AccountIdentifier::new(self_id, Some(subaccount))
+}
+
+#[update]
+async fn notify_icp_deposit(target: Principal, amount: u64) -> Result<u64, String> {
+    if amount < MIN_DEPOSIT_AMOUNT {
+        return Err(format!(
+            "Amount lower than the minimum deposit: {} ICP",
+            MIN_DEPOSIT_AMOUNT / E8S
+        ));
+    }
+    is_swap_available()?;
+    let icp_ledger = read_state(|s| s.icp_ledger_id);
+    let received_tokens = amount.checked_sub(10_000).unwrap();
```

Comment:
Nit: `10_000` literal would be more readable if it were replaced with the [DEFAULT_LEDGER_FEE constant](https://github.com/WaterNeuron/WaterNeuron/blob/1f24c7d564be33906170b7d4c2cb8cc44f7a92ec/water_neuron/src/lib.rs#L79)

Reply:
I guess amount should always be large enough to deduct fees (assuming `MIN_DEPOSIT_AMOUNT` is larger). But some explicit error handling may be useful.
```
let received_tokens = amount.checked_sub(10_000).ok_or_else(|| {
    format!("Amount too small to deduct fees: {} ICP", amount / E8S)
})?;
```